packages/core-db/src/app/queued-persisted-processor.ts

```diff
@@ -0,0 +1,330 @@
+/* External Imports */
+import {
+  bufToHexString,
+  getLogger,
+  hexStrToBuf,
+  logError,
+  Logger,
+} from '@eth-optimism/core-utils'
+
+/* Internal Imports */
+import { DB, QueuedPersistedProcessor } from '../types'
+
+const log: Logger = getLogger('base-persisted-queue')
+
+export abstract class BaseQueuedPersistedProcessor<T>
+  implements QueuedPersistedProcessor<T> {
+  public static readonly NEXT_INDEX_TO_PROCESS_KEY: string =
+    'NEXT_INDEX_TO_PROCESS'
+  public static readonly LAST_INDEX_PROCESSED: string = 'INDEX_BEING_PROCESSED'
+  public static readonly ITEM_STORAGE_KEY_PREFIX: string = 'ITEM_'
+
+  private initialized: boolean
+  private lastIndexProcessed: number
+  private nextIndexToProcess: number
+
+  protected constructor(
+    private readonly db: DB,
+    private readonly persistenceKey: string,
+    startIndex: number = 0
+  ) {
+    this.initialized = false
+    this.nextIndexToProcess = startIndex
+    this.lastIndexProcessed = startIndex - 1
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public async add(index: number, item: T): Promise<void> {
+    // Already processed this index, ignore it.
+    if (index < this.nextIndexToProcess) {
+      return
+    }
+
+    await this.persistItem(index, item)
+
+    // purposefully not awaiting response
+    this.handleIfReady(index, item)
+  }
+
+  /**
+   * @inheritDoc
+   *
+   * Note: This is not async-safe in that if it is called at the same time for the same index,
+   * the outcome is non-deterministic.
+   */
+  public async markProcessed(index: number): Promise<void> {
+    if (index <= this.lastIndexProcessed) {
+      this.log(
+        `Persisted queue received instruction to mark index ${index} processed, but last processed index is ${this.lastIndexProcessed}. Ignoring.`
+      )
+      return
+    }
+    if (index > this.lastIndexProcessed + 1) {
+      this.logError(
+        `Persisted queue [${this.persistenceKey}] received instruction to mark index ${index} as processed, but last processed index is ${this.lastIndexProcessed}. Ignoring, but this is an error!`
+      )
+      return
```

Comment:
This is a valid concern, but I'm ambivalent about it because we should never be able to _actually_ process things out of order (as unit tests demonstrate), we just may have a caller calling `markProcessed(...)` incorrectly. 

If the caller is messed up and has a bug where they are off by one in either direction, this will effectively halt, but that's more of a problem with the caller, and we at least log an error here that should in theory alert a dev about it.

That said, since this _should_ never happen, we can throw as well. This will cause our system to flap though. It'll go down, a new instance will get spun up, that will throw and go down, etc. etc.